Branch-SDK/Branch-SDK/BranchContentDiscoverer.m

```diff
@@ -0,0 +1,290 @@
+            childCount++;
+            [self discoverViewContents:view contentData:contentDataArray contentKeys:contentKeysArray clearText:isClearText ID:subViewId];
+        }
+    }
+}
+
+
+- (void)discoverFilteredViewContents:(UIViewController *)viewController contentData:(NSMutableArray *)contentDataArray contentKeys:(NSMutableArray *)contentKeysArray clearText:(BOOL)isClearText {
+    for (NSString *contentKey in contentKeysArray) {
+        NSString *contentData = [self getViewText:contentKey forController:viewController];
+        if (contentData == nil) {
+            contentData = @"";
+        }
+        if (contentDataArray) {
+            [self addFormatedContentData:contentDataArray withText:contentData clearText:isClearText];
+        }
+    }
+}
+
+- (UIView *)getRootView:(UIViewController *)viewController {
+    UIView *rootView = [viewController view];
+    if ([viewController isKindOfClass:UITableViewController.class]) {
+        rootView = ((UITableViewController *)viewController).tableView;
+    } else if ([viewController isKindOfClass:UICollectionViewController.class]) {
+        rootView = ((UICollectionViewController *)viewController).collectionView;
+    }
+    return rootView;
+}
+
+- (NSString *)getViewText:(NSString *)viewId forController:(UIViewController *)viewController {
+    NSString *viewTxt = @"";
+    if (viewController) {
+        UIView *rootView = [viewController view];
+        NSArray *viewIDSplitArray = [viewId componentsSeparatedByString:@":"];
+        if (viewIDSplitArray.count > 0) {
+            viewId = [[viewId componentsSeparatedByString:@":"] objectAtIndex:1];
+        }
+        NSArray *viewIds = [viewId componentsSeparatedByString:@"-"];
+        BOOL foundView = YES;
+        for (NSString *subViewIdStr in viewIds) {
+            NSInteger subviewId = [subViewIdStr intValue];
+            if ([rootView isKindOfClass:UITableView.class] || [rootView isKindOfClass:UICollectionView.class]) {
+                NSArray *cells = [rootView performSelector:@selector(visibleCells) withObject:nil];
+                if (cells.count > subviewId) {
+                    rootView = [cells objectAtIndex:subviewId];
+                } else {
+                    foundView = NO;
+                    break;
+                }
+            } else {
+                if ([rootView subviews].count > subviewId) {
+                    rootView = [[rootView subviews] objectAtIndex:subviewId];
+                } else {
+                    foundView = NO;
+                    break;
+                }
+            }
+        }
+        if (foundView) {
+            NSString *contentVal = [self getContentText:rootView];
+            if (contentVal) {
+                viewTxt = contentVal;
+            }
+        }
+    }
+    return viewTxt;
+}
+
+- (UIViewController *)getActiveViewController {
+    UIViewController *rootViewController = [UIApplication sharedApplication].keyWindow.rootViewController;
```

Comment:
Can you use `Class UIApplicationClass = NSClassFromString(@"UIApplication");`, then replace this line with `UIViewController *rootViewController = [UIApplicationClass sharedApplication].keyWindow.rootViewController;` We can't call UIApplication in iMessage extension. Also, sorry to add to the scope, but can you make sure the calling methods handle `nil` properly in this case?